### ISSUE TITLE
Ability to Add Custom Forms to the Browsable API

### ISSUE THREAD

**Issue by smizell** [tomchristie/django-rest-framework#4727]
## User Details

I'm building a hypermedia API that tries to expose a state machine by way of hyperlinks and forms. I'm using [Siren](https://github.com/kevinswiber/siren) for the hypermedia type and providing it by way of a custom renderer. I'm also using the browsable API to make the API accessible in the web. For Siren, I'm defining the [actions](https://github.com/kevinswiber/siren#actions-1) as DRF serializers.

## Problem

In my design, as I expose the state machine, there may be several forms for a given response. Additionally, these forms usually point to different URLs. Right now, the browsable API only shows forms that pertain to HTTP methods on response you're viewing, so this is not possible.

For example, say I was to build a task list application. Once a task was created, I may provide a form for marking the task as complete. This form may use POST on a completely separate URL from the one used to view the actual task. Once marked complete, I may provide two new forms on the task, one for archiving and one for marking as incomplete again, each of which may also have their own URLs.

## Proposed Solution

There are two things needed to make this work.

First, the `base.html` template needs to be updated with a block for the form section. This would allow me to override the forms that are provided by the browsable API's context and display my own forms. This is as simple as adding a `{% block form %}`. For now, I have copied the entire `base.html` to my applications template directory and added this block myself.

Second, I need a way to pass extra context from my API view to the renderer. This is currently only possible by passing extra properties into the `Response` and then providing a custom renderer to unpack them, attach them to the renderer instance, and then add them to the context.

```python
# Snippet from my view
if self.request.accepted_renderer.format == 'api':
    context = {'data': data, 'extra_context': extra_context}
    return Response(context, status=200)

# Custom renderer
# Note, this code breaks the OPTIONS request, but I don't have those fixes
# here for the sake of brevity.
class CustomBrowsableAPIRenderer(renderers.BrowsableAPIRenderer):
    def get_context(self, data, accepted_media_type, renderer_context):
        context = super(CustomBrowsableAPIRenderer, self).get_context(
            data, accepted_media_type, renderer_context)
        # Add to actual context
        context['extra_context'] = self.extra_context
        return context

    def render(self, context, accepted_media_type=None, renderer_context=None):
        # Attach to instance
        self.extra_context = context['extra_context']
        data = context['data']
        return super(CustomBrowsableAPIRenderer, self).render(
            data, accepted_media_type, renderer_context)
```

I would like something like `Response(data, extra_context={'foo': 'bar'})`, though this may be overloading the Response class with too much. However, the response does take a `template_name`, so it does at least have some renderer-specific arguments.

## Notes

I'd love to provide a pull request for any or all of this. If you decide to accept any of this, let me know and I'll work on a PR. I'm also happy to implement it in a way that seems best. As I've proposed, I'm asking for an `extra_context`, but I would love to do it how you see best.

Edit: last note, this solution is aimed to be general purpose. The `extra_context` may be used for things other than forms, and the `form` block would just be a way to extend that forms area.

**Comment by smizell** [comment]
Another option for my second point in the proposal would be to add something like [get_context_data](https://docs.djangoproject.com/en/1.10/topics/class-based-views/generic-display/#adding-extra-context) to the `APIView`. This however does not necessarily solve this for non-class-based views.

**Comment by tomchristie** [comment]
Hi!

There's a lot to unpack in this.

Short version of the answer is that we *do* have some hypermedia support planned for REST framework. This will tie in with the schema support in that it'll work by returning a `coreapi.Document()` instance for hypermedia responses, which a hypermedia renderer can then deal with encoding.

We could also then build in support in the browsable API for including extra controls based on links/forms present in hypermedia documents. However I think it's probably more likely that we'd look at a more comprehensive approach rather than bolting that on to the existing browsable API renderer.

Good food for thought here.

Given that you're ahead of us in the game wrt. using Siren, one option for you might be to make any changes in the form of a third party package, that we can then link to from the main docs.

A siren *renderer* could well be something that's on the cards for us to undertake, although that's a more modest proposal than also including hypermedia controls in the browsable API.

Very happy to continue the discussion here or elsewhere, although I will close off the ticket for now as we're not likely to act on the proposal in the foreseeable future, without having put some serious thought how we *fully* address hypermedia APIs.

Hope that helps! 😄 ?

**Comment by smizell** [comment]
Sounds good! I'll add a few thoughts here. Thanks for taking the time for the response 👍 

---

> without having put some serious thought how we fully address hypermedia APIs.

I want to mention, one of the needs I really have are to be able to pass extra context to the templates without conditionally doing it based on the accepted format. I don't think this is hypermedia specific, but rather just a way to pass more context to a template of any kind, not just the browsable API. The `TemplateHTMLRenderer` takes a context dict rather than a JSON-able dict, so there is a slight difference in interface for the renderers—the idea is kind of there, just not easily reusable across and entire app. That is the trick I'm currently using to bypass this, though, and I think it could be improved (however, sounds like you're already thinking about this).

If I could describe my wish for this, the `Response` would be agnostic of renderers and just take a body and headers. Then I'd like to extend upon that class with something like Django's `render` shortcut that took a template and context. It could maybe either take a `browsable` function argument or rely on a setting. Either way, I want to say, "Here is the representation of the response and here is some context to any renderer that needs it." Just thinking out loud here to share my thoughts :)

---

The other need I have is to be able to customize the browsable API template. Right now, I'm extending it change things like the title and branding via overriding the `api.html`. If I had a form block in the `base.html` file, it would basically solve this. I think you could add a content block as well and make it even more extendable.

I mention this too because I don't think this is necessarily hypermedia specific, just allowing for altering the template with Django's blocks. **Would you be interested in the idea of adding more blocks to the `base.html` file**? Happy to do a PR if so. Not trying to push here, just want to see if there are any small changes you might like that would help me out.

---

> one option for you might be to make any changes in the form of a third party package

Just as background, for this API, I'm using the [representor pattern](https://github.com/the-hypermedia-project/charter#representor-pattern) to describe my representations and then transform to other formats. One cool thing was I was able to basically take the DRF serializer and convert it directly to my representor's action. I can then transform that action to Siren, or render the serializer as an HTML form thanks to DRF's fancy `render_form` tag. It's really nice, and makes for a really cool browsable API.

I say all of that to just give some background. I think I could turn what I have into a third party package, either making it Siren-specific or representor-specific. I think it would require:

1. A wrapper for the `Response` that handled the varying behavior by media type
1. A wrapper for the browsable renderer that managed the extra context
1. Vendoring the `base.html` template because I need to change the forms area

---

Lastly, I'd love to chat about anything around hypermedia ideas if you want any additional outside input. Thanks again for the lengthy response (and of course DRF) :)

**Comment by tomchristie** [comment]
> Just as background, for this API, I'm using the representor pattern to describe my representations and then transform to other formats.

Yup, that's pretty much the same as the Core API approach. (At least as I understand it) The Document being a format agnostic representation of the resource.

Same benefit of being able to use it as a representation both on server and client side, and the ability to write generic clients using it.

As it happens I started out with Core API thinking mainly along hypermedia lines, but it works equally well as a representation for schema formats too - I've moved into focusing on that aspect more first, given the adoption, and that it can then be applied to existing APIs, as opposed to hypermedia where it needs to be designed into the API all the way through.
